routing/control_tower.go

```diff
@@ -205,82 +217,76 @@ func (p *controlTower) FetchInFlightPayments() ([]*channeldb.InFlightPayment, er
+	// Take lock before querying the db to prevent missing or duplicating an
+	// update.
+	p.paymentsMtx.Lock(paymentHash)
+	defer p.paymentsMtx.Unlock(paymentHash)
 
 	payment, err := p.db.FetchPayment(paymentHash)
 	if err != nil {
-		return false, nil, err
+		return nil, err
 	}
 
-	var event PaymentResult
+	subscriber := newControlTowerSubscriber()
 
-	switch payment.Status {
+	// Always write current payment state to the channel.
+	subscriber.queue.ChanIn() <- payment
 
-	// Payment is currently in flight. Register this subscriber and
-	// return without writing a result to the channel yet.
-	case channeldb.StatusInFlight:
+	// Payment is currently in flight. Register this subscriber for further
+	// updates. Otherwise this update is the final update and the incoming
+	// channel can be closed. This will close the queue's outgoing channel
+	// when all updates have been written.
+	if payment.Status == channeldb.StatusInFlight {
+		p.subscribersMtx.Lock()
 		p.subscribers[paymentHash] = append(
-			p.subscribers[paymentHash], c,
-		)
-
-		return true, c, nil
-
-	// Payment already succeeded. It is not necessary to register as
-	// a subscriber, because we can send the result on the channel
-	// immediately.
-	case channeldb.StatusSucceeded:
-		event = *createSuccessResult(payment.HTLCs)
-
-	// Payment already failed. It is not necessary to register as a
-	// subscriber, because we can send the result on the channel
-	// immediately.
-	case channeldb.StatusFailed:
-		event = *createFailedResult(
-			payment.HTLCs, *payment.FailureReason,
+			p.subscribers[paymentHash], subscriber,
 		)
-
-	default:
-		return false, nil, errors.New("unknown payment status")
+		p.subscribersMtx.Unlock()
+	} else {
+		close(subscriber.queue.ChanIn())
 	}
 
-	// Write immediate result to the channel.
-	c <- event
-	close(c)
-
-	return false, c, nil
+	return subscriber, nil
 }
 
-// notifyFinalEvent sends a final payment event to all subscribers of this
+// notifySubscribers sends a final payment event to all subscribers of this
 // payment. The channel will be closed after this.
-func (p *controlTower) notifyFinalEvent(paymentHash lntypes.Hash,
-	event *PaymentResult) {
+func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
+	event *channeldb.MPPayment) {
 
-	// Get all subscribers for this hash. As there is only a single outcome,
-	// the subscriber list can be cleared.
+	// Get all subscribers for this payment.
 	p.subscribersMtx.Lock()
 	list, ok := p.subscribers[paymentHash]
 	if !ok {
 		p.subscribersMtx.Unlock()
 		return
 	}
-	delete(p.subscribers, paymentHash)
 	p.subscribersMtx.Unlock()
 
-	// Notify all subscribers of the event. The subscriber channel is
-	// buffered, so it cannot block here.
+	// If the payment reached a terminal state, the subscriber list can be
+	// cleared. There won't be any more updates.
+	terminal := event.Status != channeldb.StatusInFlight
+	if terminal {
+		delete(p.subscribers, paymentHash)
```

Comment:
Access to the map without holding the mutex above. The mutex should continue to be held till the loop below.

Reply:
Oops, fixed